agent/lib_aws_sdk_php.c

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright 2024 New Relic Corporation. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/*
+ * Functions relating to instrumenting the AWS-SDK-PHP.
+ * https://github.com/aws/aws-sdk-php
+ */
+#include "php_agent.h"
+#include "php_wrapper.h"
+#include "fw_hooks.h"
+#include "fw_support.h"
+#include "util_logging.h"
+#include "php_call.h"
+#include "lib_aws_sdk_php.h"
+
+#define PHP_PACKAGE_NAME "aws/aws-sdk-php"
+#define PHP_AWS_CLASS_PREFIX "Supportability/library/aws/aws-sdk-php/"
```

Comment:
The name of this macro is very misleading - it is not a class prefix but a metrics prefix. I'm not sure what the metrics that use this prefix are supposed to capture/measure, but it looks like it not the same thing as other metrics that use `Supportability/library` prefix, which is used in [`nr_fw_support_add_library_supportability_metric`](https://github.com/newrelic/newrelic-php-agent/blob/R10.22/agent/fw_support.c#L33-L43) to indicate that a library was detected [here](https://github.com/newrelic/newrelic-php-agent/blob/R10.22/agent/php_execute.c#L921-L922). The use of this macro suggests that the intention of this metric is to capture/measure what classes/aws services? are instrumented/detected?. There already exist `Supportability/InstrumentedFunction` family of metrics with an [API to create them](https://github.com/newrelic/newrelic-php-agent/blob/R10.22/axiom/nr_txn.c#L1641-L1646) that capture/measure what functions are instrumented. If capturing/measuring what classes are instrumented is important, why not follow the same naming pattern and use the prefix `Supportability/InstrumentedClass`? But it really seems duplicated info because there are already created `Supportability/InstrumentedFunction` for all instrumented classes. If capturing/measuring what AWS services are detected is important, then `Supportability/Aws/Services/{serviceName/service client class name?}/PHP/detected` seems more appropriate as the metric name pattern and `AWS_SERVICE_DETECTED_SUPPORTABILITY_METRIC_FMT` as macro name.
```suggestion
#define AWS_SERVICE_DETECTED_SUPPORTABILITY_METRIC_FMT "Supportability/AWS/Services/%s/PHP/detected"
```

Reply:
There really is no convention about where the language should be in the supportability metric name. Some have language right after Supportablity prefix, others have them almost at the end. Because of that I don't have a strong preference one way or the other. `Supportability/AWS/Services/%s/PHP/detected` is as good as `Supportability/PHP/AWS/Services/%s/detected` to me.

Reply:
Please see: [7d98b70](https://github.com/newrelic/newrelic-php-agent/pull/932/commits/7d98b70ca2e9ada5dc79739fefa63e535ae6220d)

Instrumented function mostly did the job, but would require wrapping 50ish Aws Clients variations (and would need to determine on case by case basis what to use, for instance, dynamodbclient::__construct does not show up for PHP 8.2+).

Now we wrap the function that the aws-sdk-php uses to determine the service whenever any client is created and create a metric based off the name we snag from that.  This automatically will create metrics for all clients.

metric prefix is `Supportability/PHP/AWS/Services/%s/`
(`detected` is not required. it's only needed for frameworks to differentiate with `forced`).